### PR TITLE
feat: Adds list entity creation dialog to the lu editor

### LINKED ISSUE #7930: Add LUIS "List Entity" to LU editor toolbar

### LINKED ISSUE #7931: Add LUIS "List Entity" creation modal to LU Editor
Similar to Luis

### LINKED ISSUE #8020: NodeJS - Error sending Cards - "Error: args[0].trim is not a function"
## Describe the bug

NodeJS bot to send a respond using Cards. 
Starting the bot using Web chat results with the message "bot encountered an error or bug", the trace reviles the message "Error: Error: args[0].trim is not a function {SendActivity_Greetings}  Error occured when evaluating 'Attachments=${json(SendActivity_Greeting_attachment)}"

## Version

Version:  2.0.0
Electron: 8.2.4
Chrome: 80.0.3987.165
NodeJS: 12.13.0
V8: 8.0.426.27-electron.0

## Browser

- [ ] Chrome
- [ ] Edge

## OS

* Windows 10 Pro (20H2) x64

## To Reproduce

1. To ensure that the issue was not with my laptop setup, a clean laptop (Win10 Pro 64bit) was used. Installed NodeJS 14.17.0 and Bot Composer 2.0.0
2. Created a new NodeJS 'Empty Bot'  
3. Modified the generated Greeting "Send a response" from Text to add Attachments of 'ThumbnailCard'
4. 'Add new attachment' -> 'Create from template' -> 'Thumbnail card' - (updated required fields)
5. Started the bot and accessed it using the inbuilt Web Chat where the bot abends straightaway 
6. The error occurs for other cards 

## Expected behavior

Able to respond with cards

## Screenshots

20210607-BotComposer-NodeJs-SendActivity-ThumbnailCard-Error

## Additional context

The error occurs no matter the card type (Adaptive, Hero, etc.)

### PR DESCRIPTION
**Pull request by hatpick**
## Description

Adds list entity to LU editor toolbar
Adds list entity creation dialog to the LU editor

## Task Item

closes #7930 
closes #7931 
fixes #8020 

## Screenshots

Screen Shot 2021-06-10 at 3 30 27 PM
Screen Shot 2021-06-10 at 3 32 22 PM

### CODE REVIEW

File: Composer/packages/lib/code-editor/src/hooks/useSelection.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+import { ISelection, Selection, SelectionMode } from 'office-ui-fabric-react/lib/DetailsList';
+import * as React from 'react';
+
+export const useSelection = <T>(selectionMode?: SelectionMode) => {
+  const [items, setItems] = React.useState<T[]>([]);
+
+  const selection = React.useRef<ISelection>();
+
+  if (!selection.current) {
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/hooks/useSelection.ts:12]: Should this if-statement be in a `useEffect` hook? I guess functionally it wouldn't make a difference, but we would have to check if `selection.current` is falsy on every render.

File: Composer/packages/lib/code-editor/src/hooks/useSelection.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { ISelection, Selection, SelectionMode } from 'office-ui-fabric-react/lib/DetailsList';
+import * as React from 'react';
+
+export const useSelection = <T>(selectionMode?: SelectionMode) => {
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/hooks/useSelection.ts:7]: We should put this hook in the ui-shared package so we can use it else where in the app.

File: Composer/packages/lib/code-editor/src/lg/modalityEditors/AttachmentModalityEditor.tsx [hunk 3]
```diff
@@ -47,12 +47,40 @@ const AttachmentModalityEditor = React.memo(
         onUpdateResponseTemplate({
           Attachments: {
             kind: 'Attachments',
-            value: newItems.map((item) => `\${json(${item}())}`),
+            value: newItems.map((item) => {
+              const template = lgTemplates?.find((t) => t.name === item);
+              const isAdaptiveCard = template?.body?.includes(`"AdaptiveCard"`);
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lg/modalityEditors/AttachmentModalityEditor.tsx:52]: 👍🏻

File: Composer/packages/lib/code-editor/src/lg/modalityEditors/AttachmentModalityEditor.tsx [hunk 4]
```diff
@@ -47,12 +47,40 @@ const AttachmentModalityEditor = React.memo(
...
             valueType: 'direct',
           },
         });
       },
-      [setItems, onUpdateResponseTemplate]
+      [onUpdateResponseTemplate, lgTemplates]
+    );
+
+    const handleTemplateChange = React.useCallback(
+      (templateId: string, body?: string | undefined) => {
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lg/modalityEditors/AttachmentModalityEditor.tsx:63]:
```suggestion
      (templateId: string, body?: string) => {
```

File: Composer/packages/lib/code-editor/src/lu/DefineEntityButton.tsx [hunk 5]
```diff
@@ -48,20 +52,47 @@ const getCommandBarButton = (tooltipContent: string) =>
...
+      switch (entityType) {
+        case 'list':
+          setShowListEntityCreationDialog(true);
+          break;
+        case 'ml':
+        case 'prebuilt':
+          onDefineEntity(entityType, entityName ? { entityName } : {});
+          break;
+        default:
+          throw `${entityType} is not supported!`;
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lu/DefineEntityButton.tsx:90]: formatMessage

File: Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+const detailsRowStyles = {
+  cell: { display: 'inline-flex', alignItems: 'center' },
+  checkCell: { display: 'inline-flex', alignItems: 'center' },
+};
+const textFieldStyles = { fieldGroup: { width: '50%' } };
+const detailsListStyles = {
+  root: { overflowY: 'hidden' },
+  contentWrapper: { height: minRowHeight * defaultRowsToShow, overflowY: 'auto' },
+};
+const normalizedValueTextField = { field: { padding: 0 }, fieldGroup: { backgroundColor: 'transparent' } };
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx:58]:
```suggestion
const normalizedValueTextFieldStyles = { field: { padding: 0 }, fieldGroup: { backgroundColor: 'transparent' } };
```

File: Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  const { onDismiss, onCreateListEntity } = props;
+  const [listEntity, setListEntity] = React.useState<ListEntity>({ entityType: 'list', name: '', items: [] });
+  let listEntityId = 0;
+
+  const listRootRef = React.useRef<HTMLDivElement>(null);
+
+  const { hasErrors, nameError, itemErrors, itemsTouched } = useListEntityValidation(listEntity);
+
+  const [selectedItems, setSelectedItems] = React.useState<ListEntityItem[]>([]);
+  const selection: ISelection = React.useRef(
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx:80]: Use your fancy selection hook here.

File: Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+type Props = {
+  onCreateListEntity: (listEntity: ListEntity) => void;
+  onDismiss: () => void;
+};
+
+export const ListEntityCreationDialog = (props: Props) => {
+  const { onDismiss, onCreateListEntity } = props;
+  const [listEntity, setListEntity] = React.useState<ListEntity>({ entityType: 'list', name: '', items: [] });
+  let listEntityId = 0;
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx:73]: Won't this get reset on every render? Should you use a ref here?

File: Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+        name: formatMessage('Synonyms'),
+        isResizable: true,
+        onRender: (item: ListEntityItem) => {
+          const values = item.synonyms;
+          return (
+            <SynonymInput
+              removeOnBackspace
+              editable={false}
+              placeholder={formatMessage('Enter a synonym and press enter')}
+              tags={values}
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx:221]: No need for single use variable
```suggestion
              tags={item.synonyms}
```

File: Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+      },
+    ];
+  }, [changeNormalizedValue, changeSynonyms, itemErrors]);
+
+  const createListEntity = React.useCallback(() => {
+    onCreateListEntity(listEntity);
+  }, [onCreateListEntity, listEntity]);
+
+  const renderRow: IRenderFunction<IDetailsRowProps> = React.useCallback(
+    (rowProps) => (rowProps ? <DetailsRow {...rowProps} styles={detailsRowStyles} /> : null),
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx:236]: Will row props ever be falsy?

File: Composer/packages/lib/code-editor/src/lu/utils/entityDefinition.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    acc += `\t- ${item.normalizedValue} :${item.synonyms.length ? '\n' : ''}`;
+    if (item.synonyms.length) {
+      acc += item.synonyms.map((s) => `\t\t- ${s}`).join('\n');
+    }
+    return `${acc}\n`;
+  }, definition);
+};
+
+export const getEntityLuDefinition = (entity: Entity, entityNames: string[] = []): string => {
+  switch (entity.entityType) {
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/lu/utils/entityDefinition.ts:21]: Switch statement seems excessive for this use case

File: Composer/packages/lib/ui-shared/src/components/tags/ContentEditable.tsx [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  remove: () => void;
+  validator?: (value: string) => boolean;
+  removeOnBackspace?: boolean;
+  style?: React.CSSProperties;
+};
+
+export const ContentEditable = (props: ContentEditableProps) => {
+  const { value, className = '', style, innerEditableRef, removeOnBackspace, change, validator } = props;
+
+  let removed = false;
```
Comment (tdurnford) [inline Composer/packages/lib/ui-shared/src/components/tags/ContentEditable.tsx:27]: should these be in refs?

File: Composer/packages/lib/ui-shared/src/components/tags/TagInput.tsx [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+  removeOnBackspace?: boolean;
+};
+
+export const TagInput = (props: TagInputProps) => {
+  const { className, tags, validator, removeOnBackspace, onChange, editable, maxTags, placeholder, readOnly } = props;
+
+  const inputRef = React.createRef<HTMLInputElement>();
+  const { 0: input, 1: setInput } = React.useState('');
+
+  const onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
```
Comment (tdurnford) [inline Composer/packages/lib/ui-shared/src/components/tags/TagInput.tsx:78]: React.useCallback

File: Composer/packages/lib/ui-shared/src/components/tags/utils.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+const removeLineBreaks = (value: string) => {
+  return value.replace(/(\r\n|\n|\r)/gm, '');
+};
+
+// TAKEN FROM - https://github.com/janl/mustache.js/blob/master/mustache.js#L55
+const htmlEntityMap: Record<string, string> = {
```
Comment (tdurnford) [inline Composer/packages/lib/ui-shared/src/components/tags/utils.ts:9]: This seems useless.. Can you explain what the intent is?

**Review by tdurnford — CHANGES_REQUESTED** [review]

File: Composer/packages/lib/code-editor/src/hooks/useSelection.ts (hunk 1 above)
Comment (hatpick) [reply Composer/packages/lib/code-editor/src/hooks/useSelection.ts:12]: actually forgot to remove it, it had other issues.

File: Composer/packages/lib/code-editor/src/hooks/useSelection.ts (hunk 2 above)
Comment (hatpick) [reply Composer/packages/lib/code-editor/src/hooks/useSelection.ts:7]: removed it

File: Composer/packages/lib/code-editor/src/lu/DefineEntityButton.tsx (hunk 5 above)
Comment (hatpick) [reply Composer/packages/lib/code-editor/src/lu/DefineEntityButton.tsx:90]: it's a throw message, wont surface to UX

File: Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx (hunk 10 above)
Comment (hatpick) [reply Composer/packages/lib/code-editor/src/lu/dialogs/ListEntityCreationDialog.tsx:236]: apparently fluent library would like to think so!

File: Composer/packages/lib/ui-shared/src/components/tags/TagInput.tsx (hunk 13 above)
Comment (hatpick) [reply Composer/packages/lib/ui-shared/src/components/tags/TagInput.tsx:78]: wont make a difference

File: Composer/packages/lib/ui-shared/src/components/tags/utils.ts (hunk 14 above)
Comment (hatpick) [reply Composer/packages/lib/ui-shared/src/components/tags/utils.ts:9]: makes it easier to extend and look up is o(1)

File: Composer/packages/lib/code-editor/src/lu/utils/entityDefinition.ts (hunk 11 above)
Comment (hatpick) [reply Composer/packages/lib/code-editor/src/lu/utils/entityDefinition.ts:21]: it's extendable

**Comment by hatpick** [comment]
will create a clean one

**Pull request closed without merging**